npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -1,21 +1,193 @@
 package ipsets
 
 import (
+	"bytes"
 	"fmt"
+	"os"
+	"strings"
 
-	"github.com/Azure/azure-container-networking/npm/util/errors"
-	"k8s.io/klog"
+	"github.com/Azure/azure-container-networking/npm/util"
+	kexec "k8s.io/utils/exec"
 )
 
+// TODO look at current errors in kusto
+
+// TODO eventually, have multiple retries and start at spot based on line number in error?
+const maxRetryCount = 1
+
+type ipsetRestoreFile struct {
+	lines      []string
+	retryCount int
+}
+
+var linuxExec kexec.Interface
+
+func assertExecExists() {
+	if linuxExec == nil {
+		linuxExec = kexec.New()
+	}
+}
+
+// don't need networkID
 func (iMgr *IPSetManager) applyIPSets(networkID string) error {
-	for setName := range iMgr.dirtyCaches {
-		set, exists := iMgr.setMap[setName] // check if the Set exists
-		if !exists {
-			return errors.Errorf(errors.AppendIPSet, false, fmt.Sprintf("member ipset %s does not exist", setName))
+	// DEBUGGING)
+	fmt.Println("DIRTY CACHE")
+	fmt.Println(iMgr.dirtyCaches)
+	fmt.Println("DELETE CACHE")
+	fmt.Println(iMgr.deleteCache)
+
+	file := initIPSetRestoreFile()
+	file.create(iMgr)
+
+	// MORE DEBUGGING
+	fmt.Println("RESTORE FILE")
+	fmt.Println(file.toString(0))
+
+	assertExecExists() // TODO remove if using osexec
+	for {
+		err := file.restore(0) // TODO could retry from line that fails
+		if err == nil {
+			return nil
+		}
+		file.retryCount += 1
+		if file.retryCount >= maxRetryCount {
+			return fmt.Errorf("failed to restore ipets after %d tries with final error: %w", maxRetryCount, err)
 		}
+	}
+}
+
+func initIPSetRestoreFile() *ipsetRestoreFile {
+	return &ipsetRestoreFile{
+		lines:      make([]string, 0),
+		retryCount: 0,
+	}
+}
+
+func (file *ipsetRestoreFile) create(iMgr *IPSetManager) {
+	// need to create all sets before possibly referencing them in lists
+	for setName := range iMgr.dirtyCaches {
+		set := iMgr.setMap[setName]
+		file.createSet(set)
+		// TEMP NOTES: uses ipset create --exist
```

Comment:
How will we manager ordering ? lets say IPSETA's member is IPSETB. in current logic we add IPSETA to dirty cache first and then IPSETB. When going through dirty cache, we will first create IPSETA and then add IPSETB as IPSETA member, this can result in set not found error right ? Can we try to move all creates to the TOP and then deal with adding or deleting members ?

Reply:
I think the code is already doing what you're suggesting (createSet() only creates sets, and updateMembers() is called after all sets are created)